syntax = "proto3";

package group_a;

// Mensagem para a chamada de execute_task
message TaskRequest {
    string task_name = 1;
    int32 lamport_time = 2;
}

message TaskResponse {
    string status = 1;
    int32 lamport_time = 2;
}

// Mensagem para o algoritmo de Bully
message ElectionMessage {
    enum MessageType {
        ELECTION = 0;
        OK = 1;
        COORDINATOR = 2;
    }
    MessageType type = 1;
    int32 sender_id = 2;
    int32 lamport_time = 3;
}

message ElectionResponse {
    string status = 1;
    int32 lamport_time = 2;
}

// Mensagem de Heartbeat
message HeartbeatMessage {
    int32 sender_id = 1;
    int32 lamport_time = 2;
}

message HeartbeatResponse {
    enum Status {
        ALIVE = 0;
    }
    Status status = 1;
    int32 lamport_time = 2;
}


// Definição do serviço
service NodeAService {
    rpc ExecuteTask(TaskRequest) returns (TaskResponse) {}
    rpc HandleElection(ElectionMessage) returns (ElectionResponse) {}
    rpc Heartbeat(HeartbeatMessage) returns (HeartbeatResponse) {}
}
